File: 고예성/7월/1회차/Programmers1844_1.java
package programmers;

import java.util.*;

public class Programmers1844_1 {
	private static final int WALL = 0;
	private static final int[] dx = {1, 0, -1, 0};
	private static final int[] dy = {0, 1, 0, -1};

	public int solution(int[][] maps) {
		int row = maps.length;
		int col = maps[0].length;
		int[][] step = new int[row][col];
		Queue<Position> queue = new LinkedList<>();

		queue.offer(new Position(0, 0));
		step[0][0] = 1;
		while (!queue.isEmpty()) {
			Position cur = queue.poll();

			for (int dir = 0; dir < 4; dir++) {
				int nx = cur.x + dx[dir];
				int ny = cur.y + dy[dir];

				if (nx < 0 || nx >= row || ny < 0 || ny >= col) {
					continue;
				}
				if (maps[nx][ny] == WALL) {
					continue;
				}
				if (step[nx][ny] > 0) {
					continue;
				}

				queue.offer(new Position(nx, ny));
				step[nx][ny] = step[cur.x][cur.y] + 1;
			}
		}

		int answer = step[row - 1][col - 1];
		if (answer == 0) {
			return -1;
		}

		return answer;
	}

	private static class Position {
		public final int x;
		public final int y;

		public Position(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}
}

File: 고예성/7월/1회차/Programmers1844_1.kt
package programmers

fun main() {
    val answer = Programmers1844_1()
        .solution(
            arrayOf(
                intArrayOf(1, 0, 1, 1, 1),
                intArrayOf(1, 0, 1, 0, 1),
                intArrayOf(1, 0, 1, 1, 1),
                intArrayOf(1, 1, 1, 0, 1),
                intArrayOf(0, 0, 0, 0, 1)
            )
        )

    println("Answer: $answer")
}


class Programmers1844_1 {
    private val WALL = 0
    private val dx = intArrayOf(1, 0, -1, 0)
    private val dy = intArrayOf(0, 1, 0, -1)

    fun solution(maps: Array<IntArray>): Int {
        val row = maps.size
        val col = maps[0].size
        val step = Array(row) { IntArray(col) }
        val queue = mutableListOf<Pair<Int, Int>>()

        queue.add(Pair(0, 0))
        step[0][0] = 1
        while (queue.isNotEmpty()) {
            val curr = queue.removeFirst()

            for (dir in 0 until 4) {
                val nx = curr.first + dx[dir]
                val ny = curr.second + dy[dir]

                if (nx < 0 || ny < 0 || nx >= row || ny >= col) {
                    continue
                }
                if (maps[nx][ny] == WALL) {
                    continue
                }
                if (step[nx][ny] > 0) {
                    continue
                }

                queue.add(Pair(nx, ny))
                step[nx][ny] = step[curr.first][curr.second] + 1
            }
        }
        val answer = step[row - 1][col - 1]

        return if (answer == 0) -1 else answer
    }
}
